util/syspolicy/source/policy_reader.go

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package source
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"slices"
+	"sort"
+	"sync"
+	"time"
+
+	"tailscale.com/util/mak"
+	"tailscale.com/util/set"
+	"tailscale.com/util/syspolicy/internal/loggerx"
+	"tailscale.com/util/syspolicy/internal/metrics"
+	"tailscale.com/util/syspolicy/setting"
+)
+
+// Reader reads all configured policy settings from a given [Store].
+// It registers a change callback with the [Store] and maintains the current version
+// of the [setting.Snapshot] by lazily re-reading policy settings from the [Store]
+// whenever a new settings snapshot is requested with [Reader.GetSettings].
+// It is safe for concurrent use.
+type Reader struct {
+	store                    Store
+	origin                   *setting.Origin
+	settings                 []*setting.Definition
+	unregisterChangeNotifier func()
+	doneCh                   chan struct{} // closed when policyCache is closed.
+
+	mu         sync.RWMutex
+	closing    bool
+	upToDate   bool
+	lastPolicy *setting.Snapshot
+	sessions   set.HandleSet[*ReadingSession]
+}
+
+// newReader returns a new [Reader] that reads policy settings from a given [Store].
+// The returned reader takes ownership of the store. If the store implements [io.Closer],
+// the returned reader will close the store when it is closed.
+func newReader(store Store, origin *setting.Origin) (*Reader, error) {
```

Comment:
It is small enough when there's just one `setting.Origin`, but it all adds up (and rounds up due to alignment) when used in  each `policy.Item` / `policy.List` within a struct like this:
```go
// ExitNodes is a policy group that controls utilizing and running exit nodes.
type ExitNodes struct {
	// AllowedSuggestedExitNodes is a list of exit node IDs that restricts which
	// exit nodes are considered when generating suggestions for exit nodes.
	AllowedSuggestedExitNodes policy.List[string] ` policy:",ios,macos,windows"`
	// AllowLANAccess controls whether whether locally accessible
	// subnets should be routed directly or via the exit node.
	AllowLANAccess policy.Item[PreferenceOption] `policy:"/ExitNodeAllowLANAccess"`
	// ExitNodeID is the exit node's node id.
	// ExitNodeID takes precedence over ExitNodeIP.
	// To find the node ID, go to /api.md#device.
	ExitNodeID policy.Item[string]
	// ExitNodeIP is the exit node's IP address.
	ExitNodeIP policy.Item[string]
	// RunExitNode controls if the device acts as an exit node. Even when
	// running as an exit node, the device must be approved by a tailnet
	// administrator. Its name is slightly awkward because RunExitNodeVisibility
	// predates this option but is preserved for backwards compatibility.
	RunExitNode policy.Item[PreferenceOption] `policy:"/AdvertiseExitNode"`
}
```
I'll submit the `syspolicy/policy` package and the actual policy structures in a future PR once this and the next one merge.

But my reasoning was to maintain a single instance of `setting.Origin` for all policy settings with the same origin, and include it in the `policy.Item` and `policy.List` structs by pointer rather than by value.
We can revisit this decision in the future PR if you don't mind.